File: src/main/java/com/trustline/trustline/appuser/service/EmailServiceImpl.java
package com.trustline.trustline.appuser.service;

import com.mailersend.sdk.emails.Email;
import com.trustline.trustline.appuser.dto.EmailRequest;
import com.trustline.trustline.appuser.model.OtpModeEnum;
import com.trustline.trustline.appuser.model.VerificationModel;
import com.trustline.trustline.appuser.repository.VerificationRepository;
import com.trustline.trustline.config.exception.BadRequestException;
import com.trustline.trustline.config.exception.NotFoundException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import com.mailersend.sdk.MailerSend;
import com.mailersend.sdk.MailerSendResponse;
import com.mailersend.sdk.exceptions.MailerSendException;

import java.time.LocalDateTime;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Slf4j
public class EmailServiceImpl implements EmailService {
    private final VerificationRepository verificationRepository;
    @Value("${email.mailerSend.api-key}")
    private String apiKey;

    @Override
    public String sendMail(EmailRequest emailRequest) {
        String messageId = null;
        Email email = new Email();
        email.setFrom("Trustline Group", "admin@trustline.com.ng");
        email.addRecipient(emailRequest.getRecipientName(), emailRequest.getRecipientEmail());

        email.setSubject(emailRequest.getSubject());

        email.setHtml(emailRequest.getHtmlTemplate());

        MailerSend ms = new MailerSend();
        ms.setToken(apiKey);

        try {
            MailerSendResponse response = ms.emails().send(email);
            messageId = response.messageId;
            System.out.println(response.messageId);
        } catch (MailerSendException e) {
            log.error(e.message);
        }
        return messageId;
    }

    @Override
    public VerificationModel saveVerification(OtpModeEnum mode, String messageId, UUID userId, String otp) {
        VerificationModel verificationModel = VerificationModel.builder()
                .pin(otp)
                .mode(OtpModeEnum.EMAIL)
                .messageId(messageId)
                .userId(userId)
                .build();
       return verificationRepository.save(verificationModel);
    }

    @Override
    public Boolean verifyOtp(UUID userId, String pin) {

        VerificationModel verificationModel = verificationRepository.findByUserIdAndPin(userId, pin).orElseThrow(() -> new NotFoundException("No previous activation found for this user"));

//        TODO: check if the time has not expired, if it has exceeded an hour fail verification
        if (verificationModel.getCreatedAt().isAfter(LocalDateTime.now().plusHours(2)))
            throw new BadRequestException("Token expired, initiate another verification");
        return true;
    }


}

File: src/main/kotlin/trustline/appuser/service/EmailServiceImpl.kt
package trustline.appuser.service;

import com.resend.Resend
import com.resend.core.exception.ResendException
import com.resend.services.emails.model.SendEmailRequest
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import trustline.appuser.dto.EmailRequest
import trustline.appuser.dto.OtpModeEnum
import trustline.appuser.dto.Status
import trustline.appuser.dto.VerificationType
import trustline.appuser.model.UserModel
import trustline.appuser.repository.VerificationRepository
import trustline.notification.model.VerificationModel
import java.util.*


@Service
class EmailServiceImpl(
    private val verificationRepository: VerificationRepository,
    @Value("\${email.resend.api-key}")
    private val apiKey: String
) : EmailService {

    private val log = KotlinLogging.logger {}
    override fun sendMail(emailRequest: EmailRequest): String? {
        val resend = Resend(apiKey)

        val params = SendEmailRequest.builder()
            .from("admin@trustline.com.ng")
            .to(emailRequest.recipientEmail)
            .subject(emailRequest.subject)
            .html(emailRequest.htmlTemplate)
            .build()

        var messageId: String? = null


        try {
            val resendResponse = resend.emails().send(params)
            messageId = resendResponse.id
        } catch (e: ResendException) {
            log.error { "Unable to send email due to ${e.message}" }
        }

        return messageId
    }


    override fun saveVerification(
        mode: OtpModeEnum,
        messageId: String,
        user: UserModel,
        otp: String,
        type: VerificationType,
        status: Status
    ): VerificationModel? {

        val verificationModel =
            VerificationModel(
                pin = otp,
                mode = OtpModeEnum.EMAIL,
                messageId = messageId,
                type = type,
                user = user,
                status = status
            )

        return verificationRepository.save(verificationModel)
    }

    override fun getVerificationById(id: UUID): Optional<VerificationModel> {
        return verificationRepository.findById(id);
    }

    override fun getVerificationByIdAndStatus(id: UUID, status: Status): VerificationModel? {
        return verificationRepository.findByIdAndStatus(id, status)
    }

    override fun getbyUserIdAndPinAndStatus(userId: UUID, pin: String, status: Status): VerificationModel? {
        return verificationRepository.findByUserIdAndPinAndStatus(userId, pin, status);
    }

    override fun updateVerification(verificationModel: VerificationModel): VerificationModel? {
        return verificationRepository.save(verificationModel)
    }


}
